relayer/src/event/monitor/error.rs

```diff
@@ -0,0 +1,62 @@
+use flex_error::{define_error, TraceError};
+
+use tendermint_rpc::{Error as RpcError, Url};
+
+use ibc::core::ics24_host::identifier::ChainId;
+
+define_error! {
+    #[derive(Debug, Clone)]
+    Error {
+        WebSocketDriver
+            [ TraceError<RpcError> ]
+            |_| { "WebSocket driver failed" },
+
+        ClientCreationFailed
+            { chain_id: ChainId, address: Url }
+            |e| { format!("failed to create WebSocket driver for chain {0} with address {1}", e.chain_id, e.address) },
+
+        ClientTerminationFailed
+            [ TraceError<tokio::task::JoinError> ]
+            |_| { "failed to terminate previous WebSocket driver" },
+
+        ClientCompletionFailed
+            [ TraceError<RpcError> ]
+            |_| { "failed to run previous WebSocket driver to completion" },
+
+        ClientSubscriptionFailed
+            [ TraceError<RpcError> ]
+            |_| { "failed to run previous WebSocket driver to completion" },
+
+        NextEventBatchFailed
+            [ TraceError<RpcError> ]
+            |_| { "failed to collect events over WebSocket subscription" },
+
+        CollectEventsFailed
+            { reason: String }
+            |e| { format!("failed to extract IBC events: {0}", e.reason) },
+
+        ChannelSendFailed
+            |_| { "failed to send event batch through channel" },
+
+        SubscriptionCancelled
+            [ TraceError<RpcError> ]
+            |_| { "subscription cancelled" },
+
+        Rpc
+            [ TraceError<RpcError> ]
+            |_| { "RPC error" },
+    }
+}
+
+impl Error {
+    pub fn canceled_or_generic(e: RpcError) -> Self {
+        use tendermint_rpc::error::ErrorDetail;
+
+        match e.detail() {
+            ErrorDetail::Server(detail) if detail.reason.contains("subscription was cancelled") => {
```

Comment:
Can we demand a more type-checked way to identify this reason from `tendermint-rpc`?
Matching by string is liable to silently stopping to work, unless the tendermint-rpc developers commit to stable reason strings as part of their API.

Reply:
This error is actually emitted as a string by the Cosmos SDK, so there's not much we can do here aside from moving that very same logic into `tendermint-rpc`.